src/Test/Inspection/Core.hs

```diff
@@ -203,43 +205,56 @@ eqSlice it slice1 slice2
 -- | Returns @True@ if the given core expression mentions no type constructor
 -- anywhere that has the given name.
 freeOfType :: Slice -> [Name] -> Maybe (Var, CoreExpr)
-freeOfType slice tcNs = allTyCons (\tc -> getName tc `notElem` tcNs) slice
-
-allTyCons :: (TyCon -> Bool) -> Slice -> Maybe (Var, CoreExpr)
-allTyCons predicate slice = listToMaybe [ (v,e) | (v,e) <- slice, not (go e) ]
+freeOfType slice tcNs =
+      fmap (\(a,b,_) -> (a,b))
+    $ allTyCons (\tc -> getName tc `notElem` tcNs) slice
+
+allTyCons :: (TyCon -> Bool) -> Slice -> Maybe (Var, CoreExpr, TyCon)
+allTyCons predicate slice =
+    listToMaybe $ do
+        (v,e) <- slice
+        case runExcept (go e) of
+            Right _ -> A.empty
+            Left tc -> return (v,e,tc)
```

Comment:
Does this warrant a monad and the slightly complicated control flow of the `Except` monad?

It seems you could change the return type of `go` from `Bool` to `[TyCon]`, simply replacing `&&` with `++`; `True` with `[]`, `all` with `concatMap`

Reply:
Yes, that should work.